File: frontend/lib/models/bootie.dart
class Bootie {
  final int id;
  final String title;
  final String? description;
  final String category;
  final String status;
  final double? recommendedBounty;
  final double? finalBounty;
  final String? primaryImageUrl;
  final List<String> alternateImageUrls;
  final List<String> editedImageUrls;
  final Location? location;
  final DateTime createdAt;
  final DateTime? finalizedAt;
  final String? researchSummary;
  final String? researchReasoning;

  Bootie({
    required this.id,
    required this.title,
    this.description,
    required this.category,
    required this.status,
    this.recommendedBounty,
    this.finalBounty,
    this.primaryImageUrl,
    this.alternateImageUrls = const [],
    this.editedImageUrls = const [],
    this.location,
    required this.createdAt,
    this.finalizedAt,
    this.researchSummary,
    this.researchReasoning,
  });

  factory Bootie.fromJson(Map<String, dynamic> json) {
    return Bootie(
      id: json['id'] as int,
      title: json['title'] as String,
      description: json['description'] as String?,
      category: json['category'] as String,
      status: json['status'] as String,
      recommendedBounty: json['recommended_bounty'] != null
          ? (json['recommended_bounty'] as num).toDouble()
          : null,
      finalBounty: json['final_bounty'] != null
          ? (json['final_bounty'] as num).toDouble()
          : null,
      primaryImageUrl: json['primary_image_url'] as String?,
      alternateImageUrls: (json['alternate_image_urls'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      editedImageUrls: (json['edited_image_urls'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      location: json['location'] != null
          ? Location.fromJson(json['location'] as Map<String, dynamic>)
          : null,
      createdAt: DateTime.parse(json['created_at'] as String),
      finalizedAt: json['finalized_at'] != null
          ? DateTime.parse(json['finalized_at'] as String)
          : null,
      researchSummary: json['research_summary'] as String?,
      researchReasoning: json['research_reasoning'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'category': category,
      'status': status,
      'recommended_bounty': recommendedBounty,
      'final_bounty': finalBounty,
      'primary_image_url': primaryImageUrl,
      'alternate_image_urls': alternateImageUrls,
      'edited_image_urls': editedImageUrls,
      'location': location?.toJson(),
      'created_at': createdAt.toIso8601String(),
      'finalized_at': finalizedAt?.toIso8601String(),
      'research_summary': researchSummary,
      'research_reasoning': researchReasoning,
    };
  }

  bool get isCaptured => status == 'captured';
  bool get isSubmitted => status == 'submitted';
  bool get isResearching => status == 'researching';
  bool get isResearched => status == 'researched';
  bool get isFinalized => status == 'finalized';

  String get statusDisplayName {
    switch (status) {
      case 'captured':
        return 'Captured';
      case 'submitted':
        return 'Submitted';
      case 'researching':
        return 'Researching';
      case 'researched':
        return 'Researched';
      case 'finalized':
        return 'Finalized';
      default:
        return status;
    }
  }

  String get categoryDisplayName {
    switch (category) {
      case 'used_goods':
        return 'Used Goods';
      case 'antiques':
        return 'Antiques';
      case 'electronics':
        return 'Electronics';
      case 'collectibles':
        return 'Collectibles';
      case 'weaponry':
        return 'Weaponry';
      case 'artifacts':
        return 'Artifacts';
      case 'data_logs':
        return 'Data Logs';
      case 'miscellaneous':
        return 'Miscellaneous';
      default:
        return category;
    }
  }
}

class Location {
  final int id;
  final String name;
  final String? address;
  final String? city;
  final String? state;
  final String? zipCode;
  final String? fullAddress;
  final bool active;

  Location({
    required this.id,
    required this.name,
    this.address,
    this.city,
    this.state,
    this.zipCode,
    this.fullAddress,
    required this.active,
  });

  factory Location.fromJson(Map<String, dynamic> json) {
    return Location(
      id: json['id'] as int,
      name: json['name'] as String,
      address: json['address'] as String?,
      city: json['city'] as String?,
      state: json['state'] as String?,
      zipCode: json['zip_code'] as String?,
      fullAddress: json['full_address'] as String?,
      active: json['active'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'address': address,
      'city': city,
      'state': state,
      'zip_code': zipCode,
      'full_address': fullAddress,
      'active': active,
    };
  }
}

File: app/models/bootie.rb
# Bootie Model
#
# Represents a captured item/inventory entry in the system.
# Booties go through a workflow: captured -> submitted -> researching -> researched -> finalized
#
# A Bootie is "finalized" when it's published to the Square catalog for sale.
#
# @see PRODUCT_PROFILE.md for detailed feature descriptions
class Bootie < FirestoreModel
  # Define attributes
  attribute :title, :string
  attribute :description, :string
  attribute :category, :string
  attribute :status, :string, default: 'captured'
  attribute :user_id, :string
  attribute :location_id, :string
  attribute :primary_image_url, :string
  attribute :alternate_image_urls, default: []
  attribute :edited_image_urls, default: []
  attribute :recommended_bounty, :decimal
  attribute :final_bounty, :decimal
  attribute :research_summary, :string
  attribute :research_reasoning, :string
  attribute :research_auto_triggered, :boolean, default: false
  attribute :research_started_at, :datetime
  attribute :research_completed_at, :datetime
  attribute :square_product_id, :string
  attribute :square_variation_id, :string
  attribute :finalized_at, :datetime

  # Validations
  validates :title, presence: true
  validates :category, presence: true, inclusion: { in: %w[used_goods antiques electronics collectibles weaponry artifacts data_logs miscellaneous] }
  validates :status, presence: true, inclusion: { in: %w[captured submitted researching researched finalized] }
  validates :user_id, presence: true
  validates :location_id, presence: true

  # Scopes
  def self.by_status(status)
    where(:status, status).get
  end

  def self.by_category(category)
    where(:category, category).get
  end

  def self.by_location(location_id)
    where(:location_id, location_id).get
  end

  def self.pending_research
    # Firestore doesn't support OR queries directly, so we query each status separately
    submitted = where(:status, 'submitted').get
    researching = where(:status, 'researching').get
    submitted + researching
  end

  def self.ready_for_finalization
    where(:status, 'researched').get
  end

  def self.finalized
    where(:status, 'finalized').get
  end

  # Status helper methods
  def captured?
    status == 'captured'
  end

  def submitted?
    status == 'submitted'
  end

  def researching?
    status == 'researching'
  end

  def researched?
    status == 'researched'
  end

  def finalized?
    status == 'finalized'
  end

  # Status transition methods
  def submit!
    return false unless captured?
    update(status: 'submitted', research_auto_triggered: true)
  end

  def start_research!
    return false unless submitted?
    update(status: 'researching', research_started_at: Time.current)
  end

  def complete_research!(recommended_bounty:, research_summary:, research_reasoning:)
    return false unless researching?
    update(
      status: 'researched',
      recommended_bounty: recommended_bounty,
      research_summary: research_summary,
      research_reasoning: research_reasoning,
      research_completed_at: Time.current
    )
  end

  def finalize!(final_bounty:, square_product_id:, square_variation_id:)
    return false unless researched?
    return false if final_bounty.blank?
    
    update(
      status: 'finalized',
      final_bounty: final_bounty,
      square_product_id: square_product_id,
      square_variation_id: square_variation_id,
      finalized_at: Time.current
    )
  end

  # Associations
  def user
    User.find(user_id) if user_id.present?
  end

  def location
    Location.find(location_id) if location_id.present?
  end

  def research_logs
    ResearchLog.where(:bootie_id, id).get
  end

  def grounding_sources
    GroundingSource.where(:bootie_id, id).get
  end

  def scores
    Score.where(:bootie_id, id).get
  end
end
